[quant][core][bug fix] Corrected at::to(memory_format=...) support for quantized tensors

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75668
* #75559
* __->__ #75540
* #75711

Summary:
Previously, at::to(memory_format=....) was not properly supported for quantized tensors. e.g.,
```
In [1]: import torch
In [2]: x = torch.rand(2, 2, 2, 2)
In [3]: qx = torch.quantize_per_tensor(x, zero_point=2, dtype=torch.qint8, scale=0.5)
In [4]: qx = qx.to(memory_format=torch.channels_last)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [4], in <module>
----> 1 qx = qx.to(memory_format=torch.channels_last)

RuntimeError: false INTERNAL ASSERT FAILED at "/data/users/dzdang/pytorch/aten/src/ATen/quantized/Quantizer.cpp":430, please report a bug to PyTorch. cannot call qscheme on UnknownQuantizer
```
This occurred because of https://github.com/pytorch/pytorch/pull/75540/files#diff-d4df180433a09071e891d552426911c227b30ae9b8a8e56da31046e7ecb1afbeL79-L81 where `r` is created using `at::empty` and the information from `self`, a quantized tensor. This results in `r` being a quantized tensor, but with an `UnknownQuantizer`. 

The current code uses `at::empty_like` to create `r`, which gets routed to `empty_like_quantized`, which gives `r` the same quantizer as `self`. 

Test plan:
```
python test/test_quantization.py -k test_qtensor_to_memory_format
```

Differential Revision: [D35516418](https://our.internmc.facebook.com/intern/diff/D35516418)

dzdang committed:
[quant][core][bug fix] Corrected at::to(memory_format=...) support for quantized tensors

Summary:
Previously, at::to support for quantized tensors did not work properly,
and we had to, instead, use at::contiguous. This PR allows us to use
at::to(memory_format=...) or torch.Tensor.to(memory_format=....)
on the back- and front-ends.

Test plan:
python test/test_quantization.py -k test_qtensor_to_memory_format

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35516418).

dzdang committed:
Update on "[quant][core][bug fix] Corrected at::to(memory_format=...) support for quantized tensors"


Summary:
Previously, at::to support for quantized tensors did not work properly,
and we had to, instead, use at::contiguous. This PR allows us to use
at::to(memory_format=...) or torch.Tensor.to(memory_format=....)
on the back- and front-ends.

Test plan:
python test/test_quantization.py -k test_qtensor_to_memory_format

Differential Revision: [D35516418](https://our.internmc.facebook.com/intern/diff/D35516418)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35516418).

dzdang committed:
Update on "[quant][core][bug fix] Corrected at::to(memory_format=...) support for quantized tensors"


Summary:
Previously, at::to support for quantized tensors did not work properly,
and we had to, instead, use at::contiguous. This PR allows us to use
at::to(memory_format=...) or torch.Tensor.to(memory_format=....)
on the back- and front-ends.

Test plan:
python test/test_quantization.py -k test_qtensor_to_memory_format

Differential Revision: [D35516418](https://our.internmc.facebook.com/intern/diff/D35516418)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35516418).

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35516418).

dzdang committed:
Update on "[quant][core][bug fix] Corrected at::to(memory_format=...) support for quantized tensors"


Summary:
Previously, at::to support for quantized tensors did not work properly,
and we had to, instead, use at::contiguous. This PR allows us to use
at::to(memory_format=...) or torch.Tensor.to(memory_format=....)
on the back- and front-ends.

Test plan:
python test/test_quantization.py -k test_qtensor_to_memory_format

Differential Revision: [D35516418](https://our.internmc.facebook.com/intern/diff/D35516418)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][bug fix] Corrected at::to(memory_format=...) support for quantized tensors"


Summary:
Previously, at::to support for quantized tensors did not work properly,
and we had to, instead, use at::contiguous. This PR allows us to use
at::to(memory_format=...) or torch.Tensor.to(memory_format=....)
on the back- and front-ends.

Test plan:
python test/test_quantization.py -k test_qtensor_to_memory_format

Differential Revision: [D35516418](https://our.internmc.facebook.com/intern/diff/D35516418)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35516418).

jerryzh168 commented:
could you add a bit more detail about how this PR fixes this? it's unclear from the code

jerryzh168 commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -76,9 +76,11 @@ Tensor _to_copy(
-                     c10::nullopt);
+  // TODO: empty_quantized does not work here. It raises an exception in CheckMemoryFormat.h prior to
+  // empty_affine_quantizd/_empty_per_channel_affine_quantized calls
+  auto r = self.is_quantized() ? at::empty_like(self, memory_format)
```
can we explain the reason that we are using `empty_like` here as well? can just copy from summary

jerryzh168 commented on test/quantization/core/test_quantized_tensor.py:
```diff
@@ -140,6 +140,48 @@ def _compress_uniform_simplified(X, bit_rate, xmin, xmax, fp16_scale_bias=True):
     return Xq, loss
 
 class TestQuantizedTensor(TestCase):
+    def test_qtensor_to_memory_format(self):
```
looks like we are testing multiple things here, maybe it would make sense to split this test into multiple tests

dzdang committed:
Update on "[quant][core][bug fix] Corrected at::to(memory_format=...) support for quantized tensors"


Summary:
Previously, at::to(memory_format=....) was not properly supported for quantized tensors. e.g.,
```
In [1]: import torch
In [2]: x = torch.rand(2, 2, 2, 2)
In [3]: qx = torch.quantize_per_tensor(x, zero_point=2, dtype=torch.qint8, scale=0.5)
In [4]: qx = qx.to(memory_format=torch.channels_last)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [4], in <module>
----> 1 qx = qx.to(memory_format=torch.channels_last)

RuntimeError: false INTERNAL ASSERT FAILED at "/data/users/dzdang/pytorch/aten/src/ATen/quantized/Quantizer.cpp":430, please report a bug to PyTorch. cannot call qscheme on UnknownQuantizer
```
This occurred because of https://github.com/pytorch/pytorch/pull/75540/files#diff-d4df180433a09071e891d552426911c227b30ae9b8a8e56da31046e7ecb1afbeL79-L81 where `r` is created using `at::empty` and the information from `self`, a quantized tensor. This results in `r` being a quantized tensor, but with an `UnknownQuantizer`. 

The current code uses `at::empty_like` to create `r`, which gets routed to `empty_like_quantized`, which gives `r` the same quantizer as `self`. 

Test plan:
```
python test/test_quantization.py -k test_qtensor_to_memory_format
```

Differential Revision: [D35516418](https://our.internmc.facebook.com/intern/diff/D35516418)

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][bug fix] Corrected at::to(memory_format=...) support for quantized tensors"


Summary:
Previously, at::to(memory_format=....) was not properly supported for quantized tensors. e.g.,
```
In [1]: import torch
In [2]: x = torch.rand(2, 2, 2, 2)
In [3]: qx = torch.quantize_per_tensor(x, zero_point=2, dtype=torch.qint8, scale=0.5)
In [4]: qx = qx.to(memory_format=torch.channels_last)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [4], in <module>
----> 1 qx = qx.to(memory_format=torch.channels_last)

RuntimeError: false INTERNAL ASSERT FAILED at "/data/users/dzdang/pytorch/aten/src/ATen/quantized/Quantizer.cpp":430, please report a bug to PyTorch. cannot call qscheme on UnknownQuantizer
```
This occurred because of https://github.com/pytorch/pytorch/pull/75540/files#diff-d4df180433a09071e891d552426911c227b30ae9b8a8e56da31046e7ecb1afbeL79-L81 where `r` is created using `at::empty` and the information from `self`, a quantized tensor. This results in `r` being a quantized tensor, but with an `UnknownQuantizer`. 

The current code uses `at::empty_like` to create `r`, which gets routed to `empty_like_quantized`, which gives `r` the same quantizer as `self`. 

Test plan:
```
python test/test_quantization.py -k test_qtensor_to_memory_format
```

Differential Revision: [D35516418](https://our.internmc.facebook.com/intern/diff/D35516418)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35516418).
